pkg/api/mlflow/dao/repositories/helpers.go

```diff
@@ -0,0 +1,38 @@
+package repositories
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/G-Research/fasttrackml/pkg/api/mlflow/dao/models"
+)
+
+// makeSqlPlaceholders collects a string of (?,?,?), (?,?,?), etc
+func makeSqlPlaceholders(numberInEachSet, numberOfSets int) string {
+	placeholderArray := make([]string, numberInEachSet)
+	for i := 0; i < numberInEachSet; i++ {
+		placeholderArray[i] = "?"
+	}
+	setsArray := make([]string, numberOfSets)
+	for i := 0; i < numberOfSets; i++ {
+		setsArray[i] = fmt.Sprintf("(%s)", strings.Join(placeholderArray, ","))
+	}
+	return strings.Join(setsArray, ",")
+}
```

Comment:
this could be improved in terms of memory usage by using string builders (especially for the second iteration)

you may ask "is that not premature optimization?" and the answer to that is right up in the project name — we do care _a lot_ about optimizations in this project, and this is our main differentiator compared to the project whose API we implement

Reply:
Are not preallocated slices of the exact length efficient? The resulting arrays make it easier to use `strings.Join` to get the commas right.

Reply:
It doubles the allocated size: one for the slice of strings, and one for the joined string. It also puts more load on the GC by allocating many small strings (a slice of strings is just an array of pointers, and all of these pointer references need to be tracked by the GC)

Reply:
If you want to keep this implementation for readability, then you'd at least want to have the call to `fmt.Sprintf` done only once so that it's the same string reference that gets added n time to a slice

Reply:
so I ran a quick benchmark of this implementation with the following code:
```go
func Benchmark_makeSqlPlaceholders(b *testing.B) {
	for i := 0; i < b.N; i++ {
		makeSqlPlaceholders(1, 100)
		makeSqlPlaceholders(2, 100)
		makeSqlPlaceholders(3, 100)
	}
}
```

I get the following results:
```
Benchmark_makeSqlPlaceholders-6            51420             22895 ns/op           13892 B/op        611 allocs/op
```

When I switch to using `strings.Builder` instead, I get the following results:
```
Benchmark_makeSqlPlaceholders-6           540442              2178 ns/op            3088 B/op          6 allocs/op
```
In this particular benchmark, it is 10x faster, uses 4x less memory and does 100x less allocations.

Here is the implementation:
```go
func makeSqlPlaceholders(numberInEachSet, numberOfSets int) string {
	setLength := 2*numberInEachSet + 1
	var set strings.Builder
	set.Grow(setLength)
	set.WriteString("(")
	for i := 0; i < numberInEachSet-1; i++ {
		set.WriteString("?")
		set.WriteString(",")
	}
	set.WriteString("?)")
	setString := set.String()

	setsLength := 2*numberOfSets*setLength - 1
	var sets strings.Builder
	sets.Grow(setsLength)
	for i := 0; i < numberOfSets-1; i++ {
		sets.WriteString(setString)
		sets.WriteString(",")
	}
	sets.WriteString(set.String())
	return sets.String()
}
```

I then tried another implementation that uses `strings.Repeat` and got these results:
```
Benchmark_makeSqlPlaceholders-6           927678              1131 ns/op            3985 B/op         15 allocs/op
```
It's another 2x faster and uses ~30% more memory and 2x more allocations, but it's much simpler:
```go
func makeSqlPlaceholders(numberInEachSet, numberOfSets int) string {
	set := fmt.Sprintf("(%s)", strings.Repeat("?,", numberInEachSet-1)+"?")
	return strings.Repeat(set+",", numberOfSets-1) + set
}
```

Again, I know this is not on the critical path, but we should always strive to have great performance anywhere in this project as this is our crucial differentiator!

Reply:
okay, thanks for digging in -- I'll switch to the strings.Repeat implementation